### ISSUE TITLE
React Native

### ISSUE THREAD

**Issue by mschipperheyn** [yahoo/react-intl#119]
It seems that the library only needs a little refactoring to make it compatible for React Native. Are there any such plans?

**Comment by ericf** [comment]
Not at this time. Once React's refactor is completed to have a core `react` package that is shared between React (Web) and React Native, it'll probably make sense to support both.

**Comment by MattFoley** [comment]
+1

**Comment by MattFoley** [comment]
@mschipperheyn Do you have some notes on what changes would need to be made for this to support React Native?

**Comment by mschipperheyn** [comment]
I haven't delved that deep yet. React Native has seen some active development recently. I believe the fb has started separating out the core libraries but I'm not sure if they have reached the level of separation that Eric is waiting for

**Comment by ericf** [comment]
There's not a good story for writing a lib that works with both React and ReactNative yet. There's no abstraction of the core primitive components; e.g., `<text>` instead of `<span>`. So it's unclear what the best approach will be that doesn't make it burdensome for people to use in different React runtimes.

**Comment by mschipperheyn** [comment]
I'm not sure if creating a generic abstraction for those components is a goal of the React team. In fact, they have this made it clear at various points. So, I don't think it makes sense to wait for this. Conceivably, it might make sense to separate out the react specific tag elements. I believe most if not all functionalities of react-intl can be used as functions as well right? The tags are just syntactic sugar?

**Comment by ericf** [comment]
> I believe most if not all functionalities of react-intl can be used as functions as well right? The tags are just syntactic sugar?

Not completely… the `<FormattedMessage>` component is a great example of this. Also the components  can do optimization via `shouldComponentUpdate()` which would be difficult to do using the imperative API. I just pushed the [`v2-dev` branch](https://github.com/yahoo/react-intl/tree/v2-dev) which has lots of updates, including a new sans-mixin imperative API.

I will ping the React team to see if there's been any more progress on a pattern lib developers should follow to make things work with the different React runtimes.

**Comment by ericf** [comment]
Related: https://github.com/yahoo/react-intl/pull/131#issuecomment-133166991

**Comment by jasonmerino** [comment]
+1

**Comment by ranyefet** [comment]
Any updates? I would also love to use React-Intl in my React Native app.

**Comment by ericf** [comment]
@ranyefet how do you think we could make this work? What other libraries are you using on both web and React Native?

When I looked into this before it's the fact that React Intl renders leaf nodes that causes the problem. My plan was to wait until React 0.15 when the React core team would make this easier.

**Comment by Dindaleon** [comment]
+1

**Comment by caridy** [comment]
+100 :)

**Comment by mmeineke** [comment]
+1

**Comment by ranyefet** [comment]
@ericf I wanted to use React-Intl but I ended up using [React Native i18n](https://github.com/AlexanderZaytsev/react-native-i18n). I choose it because it was easier to setup and worked with React Native, and also uses a normal JSON to store texts which I can upload to 3rd party translation service.

My Ideal solution would be:
1. Declare localise strings within components
2. During build process, extract all strings into a JSON/.po/.mo file
3. Upload the generated file into 3rd party service, to get it translated.

Is there any plan to support the actual process of translation, or is out of scope?

Thanks!

Btw, React Native 0.18 is already uses the normal React package :)

**Comment by ngduc** [comment]
I'm planning to create a <Msg s="key"> component to abstract out messages like this:
https://github.com/ngduc/react-setup/blob/master/src/app/Home/messages.js

Maybe that component can "translate" react-intl messages to I18n.translations object to use in React Native like here:
https://github.com/AlexanderZaytsev/react-native-i18n

What do you think?

**Comment by gerbenupinion** [comment]
I got the FormattedMessage component working in react native by allowing the tagName to be a React Component or a string (html element). I also adding a props tagProps which allows you to pass through props used by the given React Component.

See https://github.com/Upinion/react-intl/commit/894ef55ffa39c4805006851ebb290661a1fec58d

```
<FormattedMessage defaultMessage='Translate this' 
id='test' 
tagName={Text} />
```

```
<FormattedMessage defaultMessage='Translate this' 
id='test' 
tagName={Text} 
tagProps={{style: { color: '#ececec'} }} />
```

It's easy to implement into the other components and would add the full react-intl functionality for react native.

**Comment by gerbenupinion** [comment]
With this commit https://github.com/Upinion/react-intl/commit/d7733cdd39c1eb69f43ae2897a19a1a3e34aa1cd the other components FormattedDate, FormattedTime, FormattedRelative, FormattedPlural and FormattedNumber work in react native as well.

**Comment by ngduc** [comment]
@gerbenupinion awesome! can't wait to see it merged. Do you have a working code with react-native to use that? Thanks.

**Comment by ericf** [comment]
@gerbenupinion I don't think using `tag` prefix makes sense, `component` is probably better. Also, `component/tagProps` feels like a leaky abstraction to me. What you can do today with no code changes is pass a function as the child to any `<Formatted*>` component to customize its rendered output.

**Comment by gerbenupinion** [comment]
I guess you encourage this kind of approach which works as well
```
<FormattedMessage defaultMessage='Translate this' 
id='test' >
    {(formattedMessage) => (
        <Text style={style: { color: '#ececec'} }>{formattedMessage}</Text>
    )}
</FormattedMessage>
```

Our problem with this method is that it requires us to include a custom function for every translation we use in react native. It adds to much complexity to the views.

@ericf What is your view at this time around supporting extras props, and changing the `<span>`? Would you mind merging this code?

I suggest renaming `tag` to `component` and `tagProps` to `componentProps`. What do you think?

**Comment by gerbenupinion** [comment]
@ngduc We got it working using the instructions of v2 and with the following code:

```
<FormattedMessage defaultMessage='Translate this' 
id='test' 
tagName={Text} 
tagProps={{style: { color: '#ececec'} }} />
```

Small note: Added `global.Intl = require('intl');` to make it work on android. See https://www.npmjs.com/package/intl.

**Comment by andpor** [comment]
Hard to follow this discussion and get a pulse of where things stand - so let me ask a question instead. 

Can the latest react-intl (v2) be used with React Native >= 0.21 or are there still open issues?

**Comment by ericf** [comment]
Haven't tried, so I don't know what the gaps still are. Which React libs that actually render UI work between React web and native?

**Comment by ericf** [comment]
The React core team is still trying to figure out how universal components would work. So I have no plans to work on this until there's a plan, but I'm happy to help contribute my thoughts to the universal components plan. 

https://github.com/reactjs/core-notes/blob/master/2016-04/april-21.md

**Comment by steida** [comment]
Check my dev stack https://github.com/este/este react-intl just works. https://github.com/este/este/blob/master/src/native/todos/NewTodo.react.js

**Comment by antoinerousseau** [comment]
You can also build React Native components using `react-intl`. I made this and it works like a charm:

```jsx
import React, {Component, PropTypes, Text} from 'react-native'
import {FormattedMessage, injectIntl, intlShape} from 'react-intl'

class Message extends Component {
  static propTypes = {
    intl: intlShape.isRequired,
    id: PropTypes.string.isRequired,
    values: PropTypes.object,
  }
  render() {
    const msg = this.props.intl.formatMessage({id: this.props.id}, this.props.values)
    return <Text>{msg}</Text>
  }
}

export default injectIntl(Message)
```

And then simply use it in your components like you would use `<FormattedMessage />`:

```jsx
import Message from './Message'

<Message id="hello" values={{user: 'World'}} />
```

**Comment by steida** [comment]
@antoinerousseau Yours proposal is very nice. Still I favour explicit over the smart code.

```jsx
<FormattedMessage {...messages.intro}>
  {message => <Text>{message}</Text>}
</FormattedMessage>
```

Little more verbose, very much explicit. We can use it for placeholder prop or any component.

**Comment by antoinerousseau** [comment]
@steida I understand but are you going to write this every time in your views when you could make your own component?

**Comment by steida** [comment]
Yes. Because we should strive for the DRY code ever possible, but not more.

![everything should be made as simple as possible but not simpler](https://cloud.githubusercontent.com/assets/66249/15916442/643cfa60-2df3-11e6-9564-ca4f663abf99.jpg)

The thing is, you never know what should be passed into Text or Input or Whatever. Abstraction sucks.

**Comment by antoinerousseau** [comment]
@steida for a lib maybe, but we are talking about final app code...

As @gerbenupinion said:
> Our problem with this method is that it requires us to include a custom function for every translation we use in react native. It adds to much complexity to the views.

**Comment by steida** [comment]
Complexity is adding another abstraction like you suggested. The solution I proposed only adds explicitness.

**Comment by joncursi** [comment]
How about `containerElement`? This is the terminology that the [material UI[(https://github.com/callemall/material-ui) team uses when you're allowed to specify a wrapper component.

**Comment by steida** [comment]
If anyone is looking for the complete workaround, check https://github.com/este/este/blob/master/src/native/app/components/formatted.js

**Comment by steida** [comment]
Btw, this is the API we need I think.

```js
// Bootstrap environment
import { setDefaultElement } from 'react-intl';
import { Text } from 'react-native';

setDefaultElement(Text);
```

**Comment by ericf** [comment]
@steida is this the way other libraries support React and React Native?

**Comment by steida** [comment]
I don't know, to be honest. But I think there are only two possible approaches:
1) export native stuff with '/native' suffix
2) set global state in Bootstrapping phase

**Comment by steida** [comment]
@ericf I think I finally figure out the simple solution. https://github.com/yahoo/react-intl/issues/600
